benchflow-experiment-manager/application/src/main/java/cloud/benchflow/experimentmanager/scheduler/ExperimentTaskScheduler.java

```diff
@@ -96,30 +97,40 @@ protected void setRunningStatesHandler(RunningStatesHandler runningStatesHandler
    */
   public synchronized void handleStartingExperiment(String experimentID) {
 
-    boolean exit = false;
     BenchFlowExperimentState experimentState;
+    BenchFlowExperimentState prevExperimentState;
 
     try {
 
       experimentState = experimentModelDAO.getExperimentState(experimentID);
+      logger.info("handleStartingExperiment: experimentState == " + experimentState);
 
-      // Stop when we reach a final starting state
-      while (!exit) {
+      while (true) {
 
-        BenchFlowExperimentState prevTestState = experimentState;
+        prevExperimentState = experimentState;
 
         try {
-          experimentState = handleExperimentState(experimentID);
+          handleExperimentState(experimentID);
         } catch (BenchFlowExperimentIDDoesNotExistException e) {
           e.printStackTrace();
           break;
         }
 
-        // Exit as soon as ready or terminated is executed
+        experimentState = experimentModelDAO.getExperimentState(experimentID);
+        logger.info("handleStartingExperiment: (while) experimentState == " + experimentState);
+        logger.info(
+            "handleStartingExperiment: (while) prevExperimentState == " + prevExperimentState);
+
+        // Exit as soon as ready or terminated has been executed
         // An experiment can go to the TERMINATED state from START in case of errors
-        if (prevTestState == experimentState
-            && (experimentState == READY || experimentState == TERMINATED)) {
-          exit = true;
+        if (prevExperimentState == READY || prevExperimentState == TERMINATED) {
+          break;
+        }
+
+        // if the dispatcher changes the state before we exit this loop we also exit
+        // TODO - add test ensuring we always exit in this case
+        if (experimentState == RUNNING) {
```

Comment:
@jeppe-style open an issue with solution where dispatcher gets the lock of the scheduler object to ensure it executes after the while loop has terminated